### DISCUSSION TITLE
Put singletons from artists without album into special directory

### DISCUSSION THREAD

**Discussion by causa-prima** [beetbox/beets#4421] [Q&A]
I am currently importing, cleaning and sanitizing my library with beets. For most artists, I have albums. However, I have a single folder for singleton tracks, in which I store the songs of artists for which I only have a few songs (imagine something like "one hit wonders" or artists of which I only like certain songs). I'd like to keep that folder in that way. I also have singleton tracks for some artists that I do have albums for, e.g. for unreleased live tracks etc. that I want to store in the artist's existing folder.

I'd expect to be able to achieve this by querying beets when importing a singleton, and checking whether the singleton's artist has some imported albums already, but I couldn't find a way to do it yet. How can I configure beets to import singletons that way (or any other way that achieves the same goal)?

(I know that, depending on the order of the import, an artist might not have an album in beets at the time of importing a singleton, but might have later. That could be handled by re-checking the whole library after importing a new file, and I think I'll be able to handle that myself)

**Comment by sampsyo** [comment]
That's a tricky one! As you mention, it's made a little harder by the fact that "has no albums" is a property that could change in the future.

I wonder if some manner of shell scripting might be the route to success here. That is, you could do `beet ls -a albumartist:$a` to check whether there are any albums associated with the artist `$a`, and if not, do `beet modify artist:$a singleton_only=1` or similar to mark the tracks by that artist. (Then you can set your `paths:` config accordingly to sort `singleton_only:1` tacks into a different directory.) You could then loop this logic over all the artists you have by doing `beet ls -f '$artist' | sort -u` or similar.

Does that make any sense?

**Reply by causa-prima** [reply]
Thanks for the ideas. I suspected I'd need to do some scripting, but I hoped there were some built-in options.

Using `beet ls -f '$artist` does list entries with collaborators on a track ("feat."). Using `$albumartists` instead does nearly work, but for some artists I still get "duplicates", e.g. because they changed their spelling at some point, an album split between two artists etc. I'll probably have to handle these cases manually anyway to put them into the same directory. I'd appreciate ideas on how to do that efficiently or how to prevent these problems.
